pkg/controller/argocd/util.go

```diff
@@ -1064,10 +1067,98 @@ func containsString(arr []string, s string) bool {
+					log.Error(err, fmt.Sprintf("failed to delete RBACs for namespace: %s", e.MetaOld.GetName()))
+				} else {
+					log.Info(fmt.Sprintf("Successfully removed the RBACs for namespace: %s", e.MetaOld.GetName()))
+				}
+			}
 			return false
 		},
 	}
 }
+
+// deleteRBACsForNamespace deletes the RBACs when the label from the namespace is removed.
+func deleteRBACsForNamespace(ownerNS, sourceNS string) error {
+	log.Info(fmt.Sprintf("Removing the RBACs created for the namespace: %s", sourceNS))
+	cfg, err := config.GetConfig()
+	if err != nil {
+		log.Error(err, "unable to get k8s config")
+		return err
+	}
+
+	k8sClient, err := kubernetes.NewForConfig(cfg)
+	if err != nil {
+		log.Error(err, "unable to create k8s client")
+		return err
+	}
+
+	// List all the roles created for ArgoCD using the label selector
+	labelSelector := metav1.LabelSelector{MatchLabels: map[string]string{common.ArgoCDKeyPartOf: common.ArgoCDAppName}}
+	roles, err := k8sClient.RbacV1().Roles(sourceNS).List(context.TODO(), metav1.ListOptions{LabelSelector: labels.Set(labelSelector.MatchLabels).String()})
+	if err != nil {
+		log.Error(err, fmt.Sprintf("failed to list roles for namespace: %s", sourceNS))
+		return err
+	}
+
+	// Delete all the retrieved roles
+	for _, role := range roles.Items {
+		err = k8sClient.RbacV1().Roles(sourceNS).Delete(context.TODO(), role.Name, metav1.DeleteOptions{})
+		if err != nil {
+			log.Error(err, fmt.Sprintf("failed to delete roles for namespace: %s", sourceNS))
+		}
+	}
+
+	// List all the roles bindings created for ArgoCD using the label selector
+	roleBindings, err := k8sClient.RbacV1().RoleBindings(sourceNS).List(context.TODO(), metav1.ListOptions{LabelSelector: labels.Set(labelSelector.MatchLabels).String()})
+	if err != nil {
+		log.Error(err, fmt.Sprintf("failed to list role bindings for namespace: %s", sourceNS))
+		return err
+	}
+
+	// Delete all the retrieved role bindings
+	for _, roleBinding := range roleBindings.Items {
+		err = k8sClient.RbacV1().RoleBindings(sourceNS).Delete(context.TODO(), roleBinding.Name, metav1.DeleteOptions{})
+		if err != nil {
+			log.Error(err, fmt.Sprintf("failed to delete role binding for namespace: %s", sourceNS))
+		}
+	}
+
+	// Get the cluster secret used for configuring ArgoCD
+	labelSelector = metav1.LabelSelector{MatchLabels: map[string]string{common.ArgoCDSecretTypeLabel: "cluster"}}
+	secrets, err := k8sClient.CoreV1().Secrets(ownerNS).List(context.TODO(), metav1.ListOptions{LabelSelector: labels.Set(labelSelector.MatchLabels).String()})
+	if err != nil {
+		log.Error(err, fmt.Sprintf("failed to retrieve secrets for namespace: %s", ownerNS))
+		return err
+	}
+	for _, secret := range secrets.Items {
+		if namespaces, ok := secret.Data["namespaces"]; ok {
+			namespaceList := strings.Split(string(namespaces), ",")
+			var result []string
+
+			for _, n := range namespaceList {
+				// remove the namespace from the list of namespaces
+				if strings.TrimSpace(n) == sourceNS {
```

Comment:
If this is the only namespace on the list and we remove it, we end up with an empty namespaces slice which means all namespaces.  Will the (non-default) argocd instance in this namespace have problem to create/sync applications in the same namespace (because it tries to access all namespaces)?

Reply:
The implementation ensures that the secret always contains the namespace where argocd instance is deployed, i.e if ArgoCD was installed in `foo` namespace, the secret would always have `foo` namespace in the list of namespaces until the ArgoCD instance is marked as a `CLUSTER_CONFIG_INSTANCE`, then it would remove the list of namespaces from the secret.

Reply:
so even if a user tries to remove the label or update the label to a different value, reconciler will make sure that the label is added again.